app/celery/research_mode_tasks.py

```diff
@@ -64,6 +74,22 @@ def aws_sns_callback(notification_id, to):
         return sns_success_callback(notification_id, destination=to, timestamp=timestamp)
 
 
+def aws_pinpoint_callback(notification_id, to):
+    now = datetime.now()
+    timestamp = now.strftime("%Y-%m-%d %H:%M:%S.%f")[:-3]
+
+    if to.strip().endswith(perm_fail):
+        return pinpoint_failed_callback(
+            "Phone is currently unreachable/unavailable", notification_id, destination=to, timestamp=timestamp
+        )
+    elif to.strip().endswith(temp_fail):
+        return pinpoint_failed_callback(
+            "Phone carrier is currently unreachable/unavailable", notification_id, destination=to, timestamp=timestamp
+        )
+    else:
+        return pinpoint_delivered_callback(notification_id, destination=to, timestamp=timestamp)
```

Comment:
It looks like only the last line is different, could we merge the logic on top into one common function, which would differ on that last function call?

Reply:
🤔 instead of
```
def aws_pinpoint_callback(notification_id, to):
    now = datetime.now()
    timestamp = now.strftime("%Y-%m-%d %H:%M:%S.%f")[:-3]

    if to.strip().endswith(perm_fail):
        return pinpoint_failed_callback(
            "Phone is currently unreachable/unavailable", notification_id, destination=to, timestamp=timestamp
        )
    elif to.strip().endswith(temp_fail):
        return pinpoint_failed_callback(
            "Phone carrier is currently unreachable/unavailable", notification_id, destination=to, timestamp=timestamp
        )
    else:
        return pinpoint_delivered_callback(notification_id, destination=to, timestamp=timestamp)
```
What about 
```
def aws_pinpoint_callback(notification_id, to):
    now = datetime.now()
    timestamp = now.strftime("%Y-%m-%d %H:%M:%S.%f")[:-3]
    using_test_perm_fail_number = to.strip().endswith(perm_fail)
    using_test_temp_fail_number = to.strip().endswith(temp_fail)

    if using_test_perm_fail_number or using_test_temp_fail_number:
        return pinpoint_failed_callback(
            "Phone is currently unreachable/unavailable" if using_test_perm_fail_number else "Phone carrier is currently unreachable/unavailable", 
            notification_id, 
            destination=to, 
            timestamp=timestamp
        )
    else:
        return pinpoint_delivered_callback(notification_id, destination=to, timestamp=timestamp)
```

Reply:
Let's ship it as is for now. I think what I was thinking of is a proper abstraction that would send the callback based on the implementation, which would simplify the call sites code, but would likely repeat in the abstractions (potentially objects and with some template patterns).